lib/launcher.js

```diff
@@ -321,38 +332,97 @@ class Launcher {
+                    // go to safe mode (or stop if already in safe mode)
+                    this.startTimes = []
+                    this.runDurations = []
+                    if (this.targetState === States.SAFE) {
+                        this.logBuffer.add({ level: 'system', msg: 'Node-RED restart loop detected whilst in safe mode. Stopping.' })
+                        this.targetState = States.STOPPED
+                    } else {
+                        this.logBuffer.add({ level: 'system', msg: 'Node-RED restart loop detected. Restarting in safe mode.' })
+                        this.targetState = States.SAFE
+                        this.start()
+                    }
+                } else {
+                    this.start()
+                }
+            } else {
+                this.start()
             }
-            return Promise.reject(new Error())
         }
 
-        setTimeout((_launcher) => {
-            // Make 15 attempts to get good response from node-red
-            //   over 3min, starting with 5s delays, ramping up to
-            //   16s starting from the 3rd retry apply a factor of
-            //   1.28 for each back off
-            // 1st TRY     :  delay 5s
-            // RETRY       :  1   2   3   4   5   6   ...  13    14
-            // DELAY (sec) :  5   5   6   8   10  13  ...  16    16
-            // TOTAL DELAY :  10  15  21  29  39  52  ...  164   180
-            promiseWithRetry(statusPoll, 14, 5000, 16000, 3, 1.28, 0.02)
-                .then(status => {
-                    _launcher.state = status
-                }).catch(() => {
-                    if (_launcher.isHealthy()) {
-                        _launcher.state = States.RUNNING
+        if (this.healthPoll) {
+            clearInterval(this.healthPoll)
+        }
+        let errorCount = 0
+        this.healthPoll = setInterval(() => {
+            if (this.state === States.STARTING || this.state === States.RUNNING) {
+                statusPoll().then(() => {
+                    // A healthly result
+                    errorCount = 0
+                    if (this.state === States.STARTING) {
+                        this.state = States.RUNNING
+                    }
+                }).catch(async _ => {
+                    // Error polling Node-RED settings page
+                    errorCount++
+                    if (errorCount === 2) {
+                        this.logBuffer.add({ level: 'system', msg: 'Node-RED hang detected.' })
+                        const targetState = this.state
+
+                        // Calling stop will clear the health poll interval
+                        await this.stop()
+                        this.targetState = targetState
+                        // Restart via restartAfterUnexpectedExit as it will check
+                        // for a restart loop
+                        restartAfterUnexpectedExit()
```

Comment:
Since `HEALTH_POLL_INTERVAL` is set to 10, this can cause any "heavy" task that takes just above 10s (e.g. a heavy loop math routine or image processing operation) to restart Node-RED.

Scenario: 
1. NR matrix math or image processing task starts 1 second before the next status poll.
2. 1 sec later, `errorCount` is incremented to 1
3. 10 sec later, `errorCount` is incremented to 2 and the process is exited. 

Is this not a bit aggressive?

Reply:
Indeed. Tuning these parameters needs some careful thought. *Any* value we pick could cause an 'expected' workload to get killed - even if locking up the event loop for long periods is an anti-pattern in node.js world.

We need to balance it against the timeliness of being able to detect a genuine hang. The longer it takes to detect and restart, the longer the flows are down and not working - impacting the availability of the instance. This will become increasingly important as we add HA failover capabilities.

Do you have an alternative set of timings to propose?

With any combination of interval/counts, there will always be a scenario where a workload appears hung to our polling - even if it isn't really. For example, a heavy tasks (that only takes 3s to run), is run every 10 seconds. It happens to align with our polling so to us, every request fails. Even with some random variance in the timing, you could just get unlucky.

Reply:
resolving this conversation following bike shedding.